File: Stage 7/ticket_search.pl
% База данных билетов
ticket(1, 1000, 5).  % Номер билета, время продажи, оставшиеся поездки
ticket(2, 2000, 2).
ticket(3, 3000, 0).
ticket(4, 1500, 7).
ticket(5, 1800, 1).

% Правило поиска билетов
find_tickets(MinNumber, MinTrips, StartTime, EndTime, TicketNumber) :-
    ticket(TicketNumber, SaleTime, RemainingTrips),
    TicketNumber > MinNumber,
    RemainingTrips >= MinTrips,
    SaleTime >= StartTime,
    SaleTime =< EndTime.

File: Stage 7/ticket_search.py
from pyswip import Prolog

class TicketSearch:
    def __init__(self, prolog_file_path):
        self.prolog = Prolog()
        self.prolog.consult(prolog_file_path)

    def search_tickets(self, min_number, min_trips, start_time, end_time):
        query = f"find_tickets({min_number}, {min_trips}, {start_time}, {end_time}, TicketNumber)"
        results = list(self.prolog.query(query))
        return [result['TicketNumber'] for result in results]